packages/web-client/app/components/card-pay/create-merchant-workflow/merchant-customization/index.ts

```diff
@@ -0,0 +1,151 @@
+    } else if (
+      this.lastCheckedMerchantId === this.merchantId &&
+      taskFor(this.validateMerchantIdTask).last?.value
+    ) {
+      return 'valid';
+    } else if (this.merchantIdValidationMessage) {
+      return 'invalid';
+    } else {
+      return 'initial';
+    }
+  }
+
+  get merchantTextColor() {
+    return mostReadable(this.merchantBgColor, [
+      '#ffffff',
+      '#000000',
+    ])!.toHexString();
+  }
+
+  @action onMerchantNameInput(value: string) {
+    this.merchantName = value;
+    this.validateMerchantName();
+  }
+
+  @action onMerchantIdInput(value: string) {
+    this.merchantId = value;
+    this.validateMerchantId();
+  }
+
+  @action onMerchantBgColorInput(event: InputEvent) {
+    let value = (event.target as HTMLInputElement).value;
+    this.merchantBgColor = value;
+  }
+
+  @action saveDetails() {
+    this.args.workflowSession.updateMany({
+      merchantName: this.merchantName,
+      merchantId: this.merchantId,
+      merchantBgColor: this.merchantBgColor,
+      merchantTextColor: this.merchantTextColor,
+    });
+    this.args.onComplete?.();
+  }
+
+  @action validateMerchantName() {
+    if (this.merchantName) {
+      this.merchantNameValidationMessage = this.merchantName.trim()
+        ? ''
+        : 'Merchant name cannot contain only spaces';
+    } else {
+      this.merchantNameValidationMessage = 'This field is required';
+    }
+  }
+
+  @action async validateMerchantId() {
+    try {
+      await taskFor(this.validateMerchantIdTask).perform();
+    } catch (e) {
+      if (didCancel(e)) {
+        return;
+      }
+      console.error(e);
+    }
+  }
+
+  @restartableTask *validateMerchantIdTask() {
+    // debounce
+    yield timeout(500);
+    let value = this.merchantId;
+
+    if (value) {
+      this.merchantIdValidationMessage = findDomainNameErrors(value);
+    } else {
+      this.merchantIdValidationMessage = 'This field is required';
+    }
+
+    if (this.merchantIdValidationMessage) {
+      return false;
+    }
+
+    // replace w call to api
+    let merchantIdExists: boolean = yield checkIfMerchantIdExists(value);
+
+    this.lastCheckedMerchantId = value;
+    if (merchantIdExists) {
+      this.merchantIdValidationMessage =
+        'This merchant ID is already taken, please choose another ID.';
+      return false;
+    }
+
+    this.merchantIdValidationMessage = '';
+    return true;
+  }
+}
+
+function findDomainNameErrors(value: string) {
+  // international domain names start with xn--
+  if (value.startsWith('xn--')) return 'Merchant ID cannot start with xn--';
+  else if (/[^a-z0-9-]/.test(value))
+    return 'Merchant ID can only contain lowercase alphabets, numbers, and hyphens';
+  else if (/^[^a-zA-Z0-9]/.test(value) || /[^a-zA-Z0-9]$/.test(value))
+    return 'Merchant ID must start and end with lowercase alphabet or number';
+  else if (value.length > 63)
+    return `Merchant ID must be at most 63 characters, currently ${value.length}`;
```

Comment:
@backspace noted yesterday that it might not be possible to allocate the full 63 characters to an ID if it was meant to be used as a subdomain, we would need to allocate `63 - domainLength` instead. 

@lukemelia do you know what the domain would be, to set this?

Reply:
Will set this to a nice round number smaller than 63 for now (50).